### PR TITLE
[dif/edn] Implement the EDN DIFs

### PR DESCRIPTION
**Pull request by mcy**
This also breaks out some common code out of the CSRNG DIF, since EDN is just a thin interface over the CSRNG.

### CODE REVIEW

File: sw/device/lib/dif/dif_edn.c [hunk 1]
```diff
@@ -4,15 +4,328 @@
...
+                                EDN_ERR_CODE_FIFO_WRITE_ERR_BIT);
+  *errors = bitfield_bit32_copy(*errors, kDifEdnErrorFifoRead, reg,
+                                EDN_ERR_CODE_FIFO_READ_ERR_BIT);
+  *errors = bitfield_bit32_copy(*errors, kDifEdnErrorFifoFullAndEmpty, reg,
+                                EDN_ERR_CODE_FIFO_STATE_ERR_BIT);
+
+  return kDifOk;
+}
+
+dif_result_t dif_edn_get_cmd_force_unhealthy_fifo(const dif_edn_t *edn,
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_edn.c:151]: Is there a reason this is split into a separate DIF, as opposed to having a single  `dif_edn_error_force()` DIF (also I think the name `dif_edn_error_force()` may be more consistent with the autogenerated `dif_edn_{alert,irq}_force()` DIFs). I.e., could you just add additional enum values to `dif_edn_error_t` so only one DIF would be needed to force all errors?

File: sw/device/lib/dif/dif_edn.c [hunk 2]
```diff
@@ -4,15 +4,328 @@
...
+    default:
+      return kDifBadArg;
+  }
+
+  DIF_RETURN_IF_ERROR(check_locked(edn));
+  mmio_region_write32(edn->base_addr, EDN_ERR_CODE_TEST_REG_OFFSET, fifo_bit);
+  return kDifOk;
+}
+
+dif_result_t dif_edn_get_cmd_force_error(const dif_edn_t *edn,
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_edn.c:173]: As mentioned above, I think the name `dif_edn_error_force()` may be more consistent with the autogenerated `dif_edn_{alert,irq}_force()` DIFs. wdyt?

File: sw/device/lib/dif/dif_edn.h [hunk 3]
```diff
@@ -83,117 +64,353 @@ typedef struct dif_edn_seed_material {
...
  * @return The result of the operation.
  */
 dif_result_t dif_edn_get_status(const dif_edn_t *edn, dif_edn_status_t flag,
                                 bool *set);
 
+/**
+ * Queries the EDN error flags.
+ *
+ * @param edn An EDN handle.
+ * @param unhealty_fifos Bitset of FIFOs in an unhealthy state; indices are
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_edn.h:258]:
```suggestion
 * @param unhealthy_fifos Bitset of FIFOs in an unhealthy state; indices are
```

File: sw/device/lib/dif/dif_edn.h [hunk 4]
```diff
@@ -83,117 +64,353 @@ typedef struct dif_edn_seed_material {
...
+ */
+OT_WARN_UNUSED_RESULT
+dif_result_t dif_edn_update(const dif_edn_t *edn,
+                            const dif_edn_seed_material_t *seed_material);
+
+/**
+ * Requests cryptographic entropy bits from the CSRNG.
+ *
+ * The prediction resistance flag as specified in SP 800-90Ar1 section
+ * 10.2.1.1 is not directily supported by the hardware. It is the
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_edn.h:356]:
```suggestion
 * 10.2.1.1 is not directly supported by the hardware. It is the
```

File: sw/device/lib/dif/BUILD [hunk 5]
```diff
@@ -192,6 +192,26 @@ cc_test(
     ],
 )
 
+cc_library(
+    name = "csrng_shared",
```
[the remark this answers is not in the document]
Comment (mcy) [reply sw/device/lib/dif/BUILD:196]: The "one lib per IP" thing is kinda silly. The shared code is actually private.

File: sw/device/lib/dif/dif_csrng_shared.c [hunk 6]
```diff
@@ -0,0 +1,36 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_csrng_shared.h"
+
+OT_WARN_UNUSED_RESULT
+dif_result_t csrng_send_app_cmd(mmio_region_t base_addr, ptrdiff_t offset,
```
[the remark this answers is not in the document]
Comment (mcy) [reply sw/device/lib/dif/dif_csrng_shared.c:8]: This isn't a DIF; it's an implementation detail. You're not supposed to include this header (which is why it's `//visibility:privatre`)

File: sw/device/lib/dif/dif_edn.h [hunk 7]
```diff
@@ -83,117 +64,353 @@ typedef struct dif_edn_seed_material {
...
+  /**
+   * Indicates a FIFO was somehow both full and empty.
+   */
+  kDifEdnErrorFifoFullAndEmpty,
+} dif_edn_error_t;
+
+/**
+ * CSRNG consume seed from entropy source enable.
+ */
+typedef enum dif_edn_entropy_src_toggle {
```
[the remark this answers is not in the document]
Comment (mcy) [reply sw/device/lib/dif/dif_edn.h:136]: CSRNG has one of these as well. I think we should decide to get rid of them in a separate PR.

File: sw/device/lib/dif/dif_edn.c (hunk 1 above)
Comment (mcy) [reply sw/device/lib/dif/dif_edn.c:151]: I've intentionally written two functions for:
- Type safety
- Consistency with CSRNG.

File: sw/device/lib/dif/dif_edn.h [hunk 8]
```diff
@@ -83,117 +64,353 @@ typedef struct dif_edn_seed_material {
...
+  kDifEdnRecoverableAlertBadFifoClear,
+  /**
+   * Indicates the genbits bus saw two identical values, indicating a possible
+   * attack.
+   */
+  kDifEdnRecoverableAlertRepeatedGenBits,
+} dif_edn_recoverable_alert_t;
+
+/**
+ * Configures EDN with runtime information.
```
[the remark this answers is not in the document]
Comment (mcy) [reply sw/device/lib/dif/dif_edn.h:185]: This is cribbed off of CSRNG. We should probably change both, together.

File: sw/device/lib/dif/dif_edn.c (hunk 2 above)
Comment (mcy) [reply sw/device/lib/dif/dif_edn.c:173]: Sure, we can do that.

File: sw/device/lib/dif/dif_csrng_shared.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (timothytrippel) [reply sw/device/lib/dif/dif_csrng_shared.c:8]: ah I missed the `//visibility:private`... my bad. marking resolved.

File: sw/device/lib/dif/BUILD (hunk 5 above)
[the remark this answers is not in the document]
Comment (timothytrippel) [reply sw/device/lib/dif/BUILD:196]: (as mentioned below), I missed the private marker. marking resolved.

File: sw/device/lib/dif/dif_edn.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (timothytrippel) [reply sw/device/lib/dif/dif_edn.h:136]: Sg, lets remove in follow up.

File: sw/device/lib/dif/dif_edn.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (timothytrippel) [reply sw/device/lib/dif/dif_edn.h:185]: same as above, a follow up is cool with me

**Review by timothytrippel — APPROVED** [review]
thanks for the fixes !

File: sw/device/lib/base/bitfield.h [hunk 9]
```diff
@@ -137,6 +137,24 @@ inline uint32_t bitfield_bit32_write(uint32_t bitfield,
                                 value ? 0x1u : 0x0u);
 }
 
+/**
+ * Copies a bit from one bit set to the other.
+ *
+ * @param dest Bitfield to write to.
+ * @param dest_bit Bit to write to.
+ * @param dest Bitfield to read from.
+ * @param dest_bit Bit to read from.
```
Comment (engdoreis) [inline sw/device/lib/base/bitfield.h:146]:
```suggestion
* @param src_bit Bit to read from.
```

File: sw/device/lib/base/bitfield.h [hunk 10]
```diff
@@ -137,6 +137,24 @@ inline uint32_t bitfield_bit32_write(uint32_t bitfield,
                                 value ? 0x1u : 0x0u);
 }
 
+/**
+ * Copies a bit from one bit set to the other.
+ *
+ * @param dest Bitfield to write to.
+ * @param dest_bit Bit to write to.
+ * @param dest Bitfield to read from.
```
Comment (engdoreis) [inline sw/device/lib/base/bitfield.h:145]:
```suggestion
* @param dest Bitfield to read from.
```

File: sw/device/lib/dif/dif_csrng.c [hunk 11]
```diff
@@ -226,57 +156,80 @@ dif_result_t dif_csrng_get_cmd_interface_status(
...
-                  kDifCsrngErrorAesSm);
-    BITSET_INSERT(status->errors, CSRNG_ERR_CODE_CMD_GEN_CNT_ERR_BIT,
-                  kDifCsrngErrorGenerateCmdCounter);
-    BITSET_INSERT(status->errors, CSRNG_ERR_CODE_FIFO_WRITE_ERR_BIT,
-                  kDifCsrngErrorFifoWrite);
-    BITSET_INSERT(status->errors, CSRNG_ERR_CODE_FIFO_READ_ERR_BIT,
-                  kDifCsrngErrorFifoRead);
-    BITSET_INSERT(status->errors, CSRNG_ERR_CODE_FIFO_STATE_ERR_BIT,
-                  kDifCsrngErrorFifoFullAndEmpty);
+    status->unhealthy_fifos =
```
Comment (engdoreis) [inline sw/device/lib/dif/dif_csrng.c:159]: Isn't worth having a look up table mapping the `kDifCsrngFifo*` to `CSRNG_ERROR_CODE_SFIFO*` and iterate through them here?

**Comment by timothytrippel** [comment]
Some chip-level tests need these so going to merge and we can address the outstanding issuses in a follow up PR.

**Pull request merged**